lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -349,15 +350,15 @@ protected function gatherScalarRowData(&$data)
 
             $fieldName = $cacheKeyInfo['fieldName'];
 
+            if (isset($cacheKeyInfo['dqlAlias'])) {
```

Comment:
What's the story with this, why was this extracted outside of if block? For fixing the bug, we are executing content of if block so shouldn't be related to this bug. Looks risky and less performant. This is very hot code path, so if we can save this isset call, we should.

Reply:
The previous code assumed that hydration info containing non-scalar data would always provide a dqlAlias, which is not the case anymore.

PHPBench report doesn't show a huge impact (ran a few times in both branches using 10revs):

**Old**

```
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
| benchmark                                                 | subject                                             | set | revs | its | mem_peak    | best          | mean          | mode          | worst         | stdev   | rstdev | diff      |
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
| SimpleInsertPerformanceBench                              | benchHydration                                      | 0   | 10   | 1   | 20,317,832b | 222,968.500μs | 222,968.500μs | 222,968.500μs | 222,968.500μs | 0.000μs | 0.00%  | 2,829.55x |
| SimpleHydrationBench                                      | benchHydration                                      | 0   | 10   | 1   | 68,035,088b | 72,636.500μs  | 72,636.500μs  | 72,636.500μs  | 72,636.500μs  | 0.000μs | 0.00%  | 921.78x   |
| MixedQueryFetchJoinFullObjectHydrationPerformanceBench    | benchHydration                                      | 0   | 10   | 1   | 17,912,192b | 5,178.100μs   | 5,178.100μs   | 5,178.100μs   | 5,178.100μs   | 0.000μs | 0.00%  | 65.71x    |
| SimpleQueryPartialObjectHydrationPerformanceBench         | benchHydration                                      | 0   | 10   | 1   | 25,619,408b | 5,007.700μs   | 5,007.700μs   | 5,007.700μs   | 5,007.700μs   | 0.000μs | 0.00%  | 63.55x    |
| SimpleQueryScalarHydrationPerformanceBench                | benchHydration                                      | 0   | 10   | 1   | 14,383,800b | 1,600.200μs   | 1,600.200μs   | 1,600.200μs   | 1,600.200μs   | 0.000μs | 0.00%  | 20.31x    |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateFixContracts                            | 0   | 10   | 1   | 6,299,880b  | 302.300μs     | 302.300μs     | 302.300μs     | 302.300μs     | 0.000μs | 0.00%  | 3.84x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateFlexContracts                           | 0   | 10   | 1   | 6,451,544b  | 514.600μs     | 514.600μs     | 514.600μs     | 514.600μs     | 0.000μs | 0.00%  | 6.53x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateUltraContracts                          | 0   | 10   | 1   | 6,318,104b  | 331.300μs     | 331.300μs     | 331.300μs     | 331.300μs     | 0.000μs | 0.00%  | 4.20x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateAllContracts                            | 0   | 10   | 1   | 6,537,920b  | 639.700μs     | 639.700μs     | 639.700μs     | 639.700μs     | 0.000μs | 0.00%  | 8.12x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertFixContracts                             | 0   | 10   | 1   | 5,912,440b  | 508.100μs     | 508.100μs     | 508.100μs     | 508.100μs     | 0.000μs | 0.00%  | 6.45x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertFlexContracts                            | 0   | 10   | 1   | 5,918,360b  | 827.100μs     | 827.100μs     | 827.100μs     | 827.100μs     | 0.000μs | 0.00%  | 10.50x    |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertUltraContracts                           | 0   | 10   | 1   | 5,963,888b  | 553.300μs     | 553.300μs     | 553.300μs     | 553.300μs     | 0.000μs | 0.00%  | 7.02x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertAllContracts                             | 0   | 10   | 1   | 6,467,976b  | 1,377.100μs   | 1,377.100μs   | 1,377.100μs   | 1,377.100μs   | 0.000μs | 0.00%  | 17.48x    |
| SimpleQueryArrayHydrationPerformanceBench                 | benchHydration                                      | 0   | 10   | 1   | 14,185,224b | 2,321.400μs   | 2,321.400μs   | 2,321.400μs   | 2,321.400μs   | 0.000μs | 0.00%  | 29.46x    |
| MixedQueryFetchJoinPartialObjectHydrationPerformanceBench | benchHydration                                      | 0   | 10   | 1   | 11,990,928b | 2,893.700μs   | 2,893.700μs   | 2,893.700μs   | 2,893.700μs   | 0.000μs | 0.00%  | 36.72x    |
| SimpleQueryFullObjectHydrationPerformanceBench            | benchHydration                                      | 0   | 10   | 1   | 62,564,816b | 14,917.200μs  | 14,917.200μs  | 14,917.200μs  | 14,917.200μs  | 0.000μs | 0.00%  | 189.30x   |
| MixedQueryFetchJoinArrayHydrationPerformanceBench         | benchHydration                                      | 0   | 10   | 1   | 31,331,656b | 4,513.500μs   | 4,513.500μs   | 4,513.500μs   | 4,513.500μs   | 0.000μs | 0.00%  | 57.28x    |
| QueryBoundParameterProcessingBench                        | benchExecuteParsedQueryWithInferredParameterType    | 0   | 10   | 1   | 5,094,376b  | 190.300μs     | 190.300μs     | 190.300μs     | 190.300μs     | 0.000μs | 0.00%  | 2.41x     |
| QueryBoundParameterProcessingBench                        | benchExecuteParsedQueryWithDeclaredParameterType    | 0   | 10   | 1   | 5,058,656b  | 78.800μs      | 78.800μs      | 78.800μs      | 78.800μs      | 0.000μs | 0.00%  | 1.00x     |
| UnitOfWorkComputeChangesBench                             | benchChangeSetComputation                           | 0   | 10   | 1   | 4,943,432b  | 795.600μs     | 795.600μs     | 795.600μs     | 795.600μs     | 0.000μs | 0.00%  | 10.10x    |
| ProxyInitializationTimeBench                              | benchCmsUserInitialization                          | 0   | 10   | 1   | 16,875,744b | 3,126.700μs   | 3,126.700μs   | 3,126.700μs   | 3,126.700μs   | 0.000μs | 0.00%  | 39.68x    |
| ProxyInitializationTimeBench                              | benchCmsEmployeeInitialization                      | 0   | 10   | 1   | 16,875,744b | 714.200μs     | 714.200μs     | 714.200μs     | 714.200μs     | 0.000μs | 0.00%  | 9.06x     |
| ProxyInitializationTimeBench                              | benchInitializationOfAlreadyInitializedCmsUsers     | 0   | 10   | 1   | 16,875,768b | 275.900μs     | 275.900μs     | 275.900μs     | 275.900μs     | 0.000μs | 0.00%  | 3.50x     |
| ProxyInitializationTimeBench                              | benchInitializationOfAlreadyInitializedCmsEmployees | 0   | 10   | 1   | 16,875,768b | 252.200μs     | 252.200μs     | 252.200μs     | 252.200μs     | 0.000μs | 0.00%  | 3.20x     |
| ProxyInstantiationTimeBench                               | benchCmsUserInstantiation                           | 0   | 10   | 1   | 4,777,016b  | 57,459.300μs  | 57,459.300μs  | 57,459.300μs  | 57,459.300μs  | 0.000μs | 0.00%  | 729.18x   |
| ProxyInstantiationTimeBench                               | benchCmsEmployeeInstantiation                       | 0   | 10   | 1   | 4,572,936b  | 38,870.800μs  | 38,870.800μs  | 38,870.800μs  | 38,870.800μs  | 0.000μs | 0.00%  | 493.28x   |
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
``` 

**New**

```
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
| benchmark                                                 | subject                                             | set | revs | its | mem_peak    | best          | mean          | mode          | worst         | stdev   | rstdev | diff      |
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
| SimpleInsertPerformanceBench                              | benchHydration                                      | 0   | 10   | 1   | 20,317,832b | 213,790.800μs | 213,790.800μs | 213,790.800μs | 213,790.800μs | 0.000μs | 0.00%  | 2,506.34x |
| SimpleHydrationBench                                      | benchHydration                                      | 0   | 10   | 1   | 68,035,088b | 71,509.600μs  | 71,509.600μs  | 71,509.600μs  | 71,509.600μs  | 0.000μs | 0.00%  | 838.33x   |
| MixedQueryFetchJoinFullObjectHydrationPerformanceBench    | benchHydration                                      | 0   | 10   | 1   | 17,983,936b | 5,020.200μs   | 5,020.200μs   | 5,020.200μs   | 5,020.200μs   | 0.000μs | 0.00%  | 58.85x    |
| SimpleQueryPartialObjectHydrationPerformanceBench         | benchHydration                                      | 0   | 10   | 1   | 25,623,976b | 4,972.500μs   | 4,972.500μs   | 4,972.500μs   | 4,972.500μs   | 0.000μs | 0.00%  | 58.29x    |
| SimpleQueryScalarHydrationPerformanceBench                | benchHydration                                      | 0   | 10   | 1   | 14,388,368b | 1,671.000μs   | 1,671.000μs   | 1,671.000μs   | 1,671.000μs   | 0.000μs | 0.00%  | 19.59x    |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateFixContracts                            | 0   | 10   | 1   | 6,301,208b  | 314.000μs     | 314.000μs     | 314.000μs     | 314.000μs     | 0.000μs | 0.00%  | 3.68x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateFlexContracts                           | 0   | 10   | 1   | 6,456,112b  | 465.200μs     | 465.200μs     | 465.200μs     | 465.200μs     | 0.000μs | 0.00%  | 5.45x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateUltraContracts                          | 0   | 10   | 1   | 6,322,672b  | 323.400μs     | 323.400μs     | 323.400μs     | 323.400μs     | 0.000μs | 0.00%  | 3.79x     |
| SingleTableInheritanceHydrationPerformanceBench           | benchHydrateAllContracts                            | 0   | 10   | 1   | 6,542,488b  | 747.300μs     | 747.300μs     | 747.300μs     | 747.300μs     | 0.000μs | 0.00%  | 8.76x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertFixContracts                             | 0   | 10   | 1   | 5,912,440b  | 510.400μs     | 510.400μs     | 510.400μs     | 510.400μs     | 0.000μs | 0.00%  | 5.98x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertFlexContracts                            | 0   | 10   | 1   | 5,918,360b  | 692.600μs     | 692.600μs     | 692.600μs     | 692.600μs     | 0.000μs | 0.00%  | 8.12x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertUltraContracts                           | 0   | 10   | 1   | 5,963,888b  | 544.800μs     | 544.800μs     | 544.800μs     | 544.800μs     | 0.000μs | 0.00%  | 6.39x     |
| SingleTableInheritanceInsertPerformanceBench              | benchInsertAllContracts                             | 0   | 10   | 1   | 6,467,976b  | 1,221.700μs   | 1,221.700μs   | 1,221.700μs   | 1,221.700μs   | 0.000μs | 0.00%  | 14.32x    |
| SimpleQueryArrayHydrationPerformanceBench                 | benchHydration                                      | 0   | 10   | 1   | 14,189,792b | 2,321.000μs   | 2,321.000μs   | 2,321.000μs   | 2,321.000μs   | 0.000μs | 0.00%  | 27.21x    |
| MixedQueryFetchJoinPartialObjectHydrationPerformanceBench | benchHydration                                      | 0   | 10   | 1   | 12,062,672b | 2,849.400μs   | 2,849.400μs   | 2,849.400μs   | 2,849.400μs   | 0.000μs | 0.00%  | 33.40x    |
| SimpleQueryFullObjectHydrationPerformanceBench            | benchHydration                                      | 0   | 10   | 1   | 62,569,384b | 14,219.000μs  | 14,219.000μs  | 14,219.000μs  | 14,219.000μs  | 0.000μs | 0.00%  | 166.69x   |
| MixedQueryFetchJoinArrayHydrationPerformanceBench         | benchHydration                                      | 0   | 10   | 1   | 31,403,888b | 4,522.500μs   | 4,522.500μs   | 4,522.500μs   | 4,522.500μs   | 0.000μs | 0.00%  | 53.02x    |
| QueryBoundParameterProcessingBench                        | benchExecuteParsedQueryWithInferredParameterType    | 0   | 10   | 1   | 5,094,376b  | 186.500μs     | 186.500μs     | 186.500μs     | 186.500μs     | 0.000μs | 0.00%  | 2.19x     |
| QueryBoundParameterProcessingBench                        | benchExecuteParsedQueryWithDeclaredParameterType    | 0   | 10   | 1   | 5,058,656b  | 85.300μs      | 85.300μs      | 85.300μs      | 85.300μs      | 0.000μs | 0.00%  | 1.00x     |
| UnitOfWorkComputeChangesBench                             | benchChangeSetComputation                           | 0   | 10   | 1   | 4,943,432b  | 620.600μs     | 620.600μs     | 620.600μs     | 620.600μs     | 0.000μs | 0.00%  | 7.28x     |
| ProxyInitializationTimeBench                              | benchCmsUserInitialization                          | 0   | 10   | 1   | 16,875,744b | 2,962.000μs   | 2,962.000μs   | 2,962.000μs   | 2,962.000μs   | 0.000μs | 0.00%  | 34.72x    |
| ProxyInitializationTimeBench                              | benchCmsEmployeeInitialization                      | 0   | 10   | 1   | 16,875,744b | 717.000μs     | 717.000μs     | 717.000μs     | 717.000μs     | 0.000μs | 0.00%  | 8.41x     |
| ProxyInitializationTimeBench                              | benchInitializationOfAlreadyInitializedCmsUsers     | 0   | 10   | 1   | 16,875,768b | 263.400μs     | 263.400μs     | 263.400μs     | 263.400μs     | 0.000μs | 0.00%  | 3.09x     |
| ProxyInitializationTimeBench                              | benchInitializationOfAlreadyInitializedCmsEmployees | 0   | 10   | 1   | 16,875,768b | 262.600μs     | 262.600μs     | 262.600μs     | 262.600μs     | 0.000μs | 0.00%  | 3.08x     |
| ProxyInstantiationTimeBench                               | benchCmsUserInstantiation                           | 0   | 10   | 1   | 4,777,016b  | 55,764.200μs  | 55,764.200μs  | 55,764.200μs  | 55,764.200μs  | 0.000μs | 0.00%  | 653.74x   |
| ProxyInstantiationTimeBench                               | benchCmsEmployeeInstantiation                       | 0   | 10   | 1   | 4,572,936b  | 37,987.000μs  | 37,987.000μs  | 37,987.000μs  | 37,987.000μs  | 0.000μs | 0.00%  | 445.33x   |
+-----------------------------------------------------------+-----------------------------------------------------+-----+------+-----+-------------+---------------+---------------+---------------+---------------+---------+--------+-----------+
```

It can be my machine too, though :)

Reply:
I didn't expect significant impact, but it's still negative impact for questionanble gain. It doesn't need to be outside of if block below, because there is no switch branch returning both `dqlAlias` and `isScalar` at the same time in `hydrateColumnInfo`. So any time `dqlAlias` is defined, if block below *will* be executed. I known people like @guilhermeblanco spent lot of time on making these tiny improvements. Lot of them combined together are significant.

Reply:
Wait... why `dqlAlias` is not defined anymore?

Reply:
Sorry, but there is always supposed to exist a `dqlAlias` in scalar scenarios. In the case there isn't an easy one (or it's a computed value, we use the `dctrn__` prefix and generate one on the fly.

Aside from that, there're a couple fixes I've made on this hydrator, so it might be worth pulling them from master to improve proper DBAL type conversion.

Reply:
Unfortunately it doesn't seem there are master changes regarding this

Reply:
> Wait... why `dqlAlias` is not defined anymore?
> Sorry, but there is always supposed to exist a `dqlAlias` in scalar scenarios. In the case there isn't an easy one (or it's a computed value, we use the `dctrn__` prefix and generate one on the fly.

There's no `dqlAlias` defined for scalar scenarios - even on `master`: https://github.com/doctrine/orm/blob/6c7bb5a27360e1293fc868b76067bc2085239d2d/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php#L402-L407

> I didn't expect significant impact, but it's still negative impact for questionanble gain. It doesn't need to be outside of if block below, because there is no switch branch returning both dqlAlias and isScalar at the same time in hydrateColumnInfo. So any time dqlAlias is defined, if block below will be executed. I known people like @guilhermeblanco spent lot of time on making these tiny improvements. Lot of them combined together are significant.

I'm quite aware of the efforts, but what you're missing here is that scalar scenarios (old and the introduced one) **don't specify a `dqlAlias`**. This patch introduces the scenario where **both `isScalar` and `dqlAlias`** aren't defined.

For this specific scenario, it's safe to add anything as `dqlAlias` because we don't care about the column names (`array_map('current', $foundIdRows)`  is used), so I'll do that to get this moving.

> Aside from that, there're a couple fixes I've made on this hydrator, so it might be worth pulling them from master to improve proper DBAL type conversion.

I wouldn't release that in a patch, we can totally do it for v2.8